### PR TITLE
🔧 avoid the installation of boost cmake config.

### PR DESCRIPTION
**Pull request by daminetreg**
* I've followed [this guide](https://hunter.readthedocs.io/en/latest/creating-new/update.html)
  step by step carefully. **[Yes]**

* I've tested this package remotely and have excluded all broken builds.
  Here is the links to the Travis/AppVeyor with status **"All passed :wink:"**:

  * https://ci.appveyor.com/project/daminetreg/hunter/builds/30255155
  * https://travis-ci.com/nxxm/hunter/builds/145347045

Dear @ruslo, @dan-42, @rbsheth,

This fixes the issues discovered in #133 : when linking to a Boost library, the Boost::boost INTERFACE lib gets to link to the `/Source` directory that hunters generate during the build.

This is due of the BoostConfig-*.cmake files generated by Boost, because they aren't cleaned up anymore by your script : `@HUNTER_SELF@/scripts/clean-boost-configs.cmake`. But we don't need it anymore as we can pass `b2 --no-cmake-config`.

The cleanup didn't work anymore is due to a change in the names, at least on macOS I could check that, that the boost b2 generated cmake package configs were named lowercase like `boost_headers-1.72.0/boost_headers-config.cmake`, and I think at least in some case takes precedence over the cmake module for finding boost. 

Presumably that's the reason why tests didn't fail when @dan-42 updated the boost version, or there is some cmakecache thing or ci cache thing involved.

Little question should I update the Boost revision to avoid any successful build because of a cache that doesn't notice the changes ?

### CODE REVIEW

**Comment by daminetreg** [comment]
All tests passed for pkg.boost. ✔️

**Comment by BenWhetton** [comment]
Does the ID value on the last line of `cmake/projects/Boost/hunter.cmake` need to be incremented before this is merged in?
See https://github.com/cpp-pm/hunter/blob/master/docs/reference/user-modules/hunter_download.rst#hunter_download

**Comment by daminetreg** [comment]
> Does the ID value on the last line of `cmake/projects/Boost/hunter.cmake` need to be incremented before this is merged in?
> See https://github.com/cpp-pm/hunter/blob/master/docs/reference/user-modules/hunter_download.rst#hunter_download

Yes that was the bit I was asking. I’ll push that change to PACKAGE_INTERNAL_DEPS_ID.

**Comment by Bjoe** [comment]
Read also my https://github.com/cpp-pm/hunter/issues/133#issuecomment-576590478 about `Boost::boost` ....

**Comment by daminetreg** [comment]
I upgraded Boost PACKAGE_INTERNAL_DEPS_ID as pointed out by @BenWhetton. And relaunched tests on the test.boost branch.  

See builds in progress : 
* https://ci.appveyor.com/project/daminetreg/hunter/builds/30255155
* https://travis-ci.com/nxxm/hunter/builds/145347045

> Read also my [#133 (comment)](https://github.com/cpp-pm/hunter/issues/133#issuecomment-576590478) about `Boost::boost` ....

I've replied here, I think you are right on the long term, but as of Boost 1.72.0 I don't see another option as this one for now.

**Comment by daminetreg** [comment]
Again, with the PACKAGE_INTERNAL_DEPS_ID updated all tests passed for test.boost. ✔️

**Comment by dan-42** [comment]
PING: @bkotzz @rbsheth 👋

**Comment by BenWhetton** [comment]
@rbsheth: Will you be able to review this soon? We are waiting for this fix to land but if it is going to take much longer we'll have to implement a local workaround for the issues we're encountering (see #141).

**Comment by Bjoe** [comment]
@rbsheth  We should update `PACKAGE_INTERNAL_DEPS_ID` also in all boost components.

If we not do this, then every hunter user they already build boost 1.72.0 from hunter version 0.23.245 must nuked `~/.hunter` directory, because for hunter the boost version is not changed so they will not build new boost package .... see also [the comment/post](https://github.com/cpp-pm/hunter/issues/141#issuecomment-581793449) in issue #141

**Comment by BenWhetton** [comment]
@daminetreg do you think you'll have time to update this soon?

---

> As @Bjoe said, the `iOS` build may need to be updated with the `--no-cmake-config` also. See https://github.com/cpp-pm/hunter/blob/master/cmake/projects/Boost/schemes/url_sha1_boost_ios_library.cmake.in
> Will adding the flag `--no-cmake-config` break older Boost versions that don't support the flag?

Are there any regression tests which cover this?

**Comment by daminetreg** [comment]
> @daminetreg do you think you'll have time to update this soon?

Yes I'll do that this Friday morning. I have a little bit of time.

> 
> > As @Bjoe said, the `iOS` build may need to be updated with the `--no-cmake-config` also. See https://github.com/cpp-pm/hunter/blob/master/cmake/projects/Boost/schemes/url_sha1_boost_ios_library.cmake.in
> > Will adding the flag `--no-cmake-config` break older Boost versions that don't support the flag?
> 
> Are there any regression tests which cover this?

Well if you want previous version to continue to work without this I need to restore the files responsible of deleting BoostConfig.cmake. I'll do something based on the version of boost if it's greater or equal 1.72 I'll do what I have here and otherwise go back.

**Comment by rbsheth** [comment]
@daminetreg Did you have an opportunity to fix this?

**Comment by madmongo1** [comment]
@rbsheth looks like @daminetreg is busy. This PR is pretty important. Boost is central to many c++ projects.
If you can offer guidance as to what remains to be done, I'd be happy to open a new PR which incorportates these changes and the final fix.

**Comment by daminetreg** [comment]
I did lag a bit, but it is almost done, this evening pr will be updated with the little bits missing to prevent hunter cache issues.

**Comment by daminetreg** [comment]
* **IMHO I'm done with all comments from the review**, and avoided changing behaviour for older boost version (tested also with an older one locally).

* Here is the links to the Travis/AppVeyor with status **"Build in Progress"**:

  * https://ci.appveyor.com/project/daminetreg/hunter/builds/30799737
  * https://travis-ci.com/nxxm/hunter/builds/148937865

**Comment by daminetreg** [comment]
For a summary: 
* Rebased on master to integrate the changes made in between
* Updated the ios build
* Implemented the use of bjam --no-cmake-configs only if version is >= 1.72.0.
* Updated all PACKAGE_INTERNAL_DEPS_ID

**Comment by madmongo1** [comment]
Thank you for taking the time to do this. I bow before your CMake expertise!

**Comment by daminetreg** [comment]
> Thank you for taking the time to do this. I bow before your CMake expertise!

Thank you very much 😄 

However the CI saw I did oversee the fact that the generator expression didn't work on all platform the same, and before all I used them in a context where they are not officially allowed by the documentation. So I removed them in favor of a bit more traditional approach.

This has the downside to try to clean always the boost config cmakes even when we don't install them, but it's safer as it is not allowed by the cmake documentation (even though this part worked everywhere due to the fact the file is generated by `file(GENERATE OUTPUT ...)`.)

And to that while they came quite handy in this situations they are not thought for the kind of use I did, so better less clever but more solid.

* Here is the new links to the Travis/AppVeyor with status **"Build in Progress"**:  
  * https://ci.appveyor.com/project/daminetreg/hunter/builds/30809619
  * https://travis-ci.com/nxxm/hunter/builds/149020624

File: cmake/projects/Boost/schemes/url_sha1_boost.cmake.in [hunk 1]
```diff
@@ -103,6 +103,12 @@ configure_file(
   @ONLY
 )
 
+
+set(pass_boost_no_cmake_config )
+if (@HUNTER_Boost_VERSION@ VERSION_GREATER_EQUAL 1.72.0)
```
Comment (rbsheth) [inline cmake/projects/Boost/schemes/url_sha1_boost.cmake.in]: Will this work when the version has a suffix like `-p0`? If tests are passing, I guess so?

Comment (daminetreg) [reply cmake/projects/Boost/schemes/url_sha1_boost.cmake.in]: Yes it does work with -p0 it ends capturing version numbers as soon as something else than . and numbers are found. So the -p0 just get ignored.

Comment (rbsheth) [reply cmake/projects/Boost/schemes/url_sha1_boost.cmake.in]: Sweet

**Comment by daminetreg** [comment]
Thank you very much for the review. Tests are all green and passed.

By the way where is @ruslo ?

**Comment by rbsheth** [comment]
@daminetreg Great! @ruslo has passed this project onto myself and @bkotzz - he is focusing on other things.

**Comment by daminetreg** [comment]
Thanks for merging ^^ Hope it will work for the users as CI confirmed.

@ruslo is a terribly great expert in Cmake, hope he can comeback someday ^^.

@rbsheth keep up with the super good work !

**Comment by madmongo1** [comment]
confirmed working here:
https://github.com/madmongo1/dynamic_buffer_v2

**Pull request merged**